supplier/models.py

```diff
@@ -0,0 +1,121 @@
+    password = models.CharField(max_length=16,
+                                validators=[MinLengthValidator(6)],
+                                null=False,
+                                help_text="Password")
+    user_type = models.CharField(max_length=16,
+                                 validators=[MinLengthValidator(6)],
+                                 null=False,
+                                 choices=[('client', 'client'), ('supplier', 'supplier')],
+                                 help_text="User Type - supplier/client")
+
+    class Meta:
+        abstract = True
+
+
+class Supplier(User):
+    """Supplier class inherit from User abstract class.
+
+    Supplier class lets the app user to do variety of actions
+    which evolves the Product and Delivery class.
+
+    Attributes:
+        business_name: a string which holds the Supplier business name.
+    """
+    business_name = models.CharField(max_length=32,
+                                     validators=[MinLengthValidator(6)],
+                                     null=False,
+                                     help_text="Business Name")
+
+    @staticmethod
+    def filter_by_field(field, field_value):
+        """Fetches list of Suppliers filtered with given field and given value.
+
+        Args:
+        field:
+            Supplier attribute.
+        field_value:
+            value which will filter the suppliers.
+
+        Returns:
+            returns a queryset containing the objects that match the specified filter conditions.
+
+        Raises:
+        FieldDoesNotExist error: if Supplier doesn't have the specified attribute.
+        """
+        if hasattr(Supplier, str(field)):
+            return Supplier.objects.filter(**{field: field_value})
+        raise FieldDoesNotExist
+
+    @staticmethod
+    def get_all_suppliers():
+        """Fetches list of all Suppliers.
+
+        Args:
+            None.
+
+        Returns:
+            returns a queryset containing all the Suppliers.
+        """
+        return Supplier.objects.all()
+
+    @staticmethod
+    def delete_supplier(supplier):
+        """Deletes supplier 'safely'
+
+        Args:
+            supplier - the Supplier wanted to be deleted
+
+        Returns:
+            None.
+
+        raises:
+        ObjectDoesNotExists error: if the supplier is not in DB.
+        """
+        if supplier not in Supplier.get_all_suppliers():
+            raise ObjectDoesNotExist
+        supplier.delete()
+
+    @staticmethod
+    def save_supplier(supplier):
+        """Saves new supplier 'safely'
+
+        Args:
+            supplier - the Supplier wanted to be saved.
+
+        Returns:
+            None.
+
+        raises:
+        ValidationError error: if fields input aren't valid.
+        """
+        try:
+            Supplier.full_clean(supplier)
+            supplier.save()
+        except ValidationError as error:
+            raise error
```

Comment:
Is there a need for this method? can't we just use the builtin `supplier.save()`?
what is the point of the exception handling if we don't really handle the exception? (we catch it only to throw it)

Reply:
if ill just use save method it wont throw an error, the method full clean throws validation error, my logic is this:
try to check for right input(full clean method) - > if it throws the user should catch it, otherwise it would save the instance in the DB.
i could do the code with out try catch block, but i saw it as more readable with raising the error.

Reply:
But the user does not catch it, it raises it immediately after it does; so not only it does not save it in the DB, but it also crashes the program. As a user, I would probably expect some kind of warning that'll help me understand what I did wrong. 

```
You should only need to call a model’s [full_clean()](https://docs.djangoproject.com/en/4.1/ref/models/instances/#django.db.models.Model.full_clean) method if you plan to handle validation errors yourself, or if you have excluded fields from the [ModelForm](https://docs.djangoproject.com/en/4.1/topics/forms/modelforms/#django.forms.ModelForm) that require validation.
```

Reply:
isnt that the user responsibility to catch it? im kinda confused here.
there is a need to this test since the method save() doesnt check for the fields, since that, to save a Supplier full clean method is needed.
full clean method raises the validation error, the logic behind my code is to make it more understandable, also mentioned it in the docstring, as i see it, the minimum a user could do is check the doc string to understand what crashes its program.

Reply:
fixed